cmd/kubeadm/app/apis/kubeadm/v1alpha1/upgrade.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"bytes"
+	"errors"
+	"reflect"
+
+	"github.com/json-iterator/go"
+	"github.com/ugorji/go/codec"
+	yaml "gopkg.in/yaml.v2"
+
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+)
+
+var json = jsoniter.ConfigCompatibleWithStandardLibrary
+
+type MigrationStep func(map[string]interface{}) error
+
+var Migrations = map[string][]MigrationStep{
```

Comment:
I think that it is necessary to link migration steps to a version, and then identify the steps to execute according to current/target version when executing the `Migrate` method. e.g. 
- ProxyFeatureListToMap linked v1.10.0
And so 
- upgrade from v1.9.3 to v1.9.6:  no migration steps >v1.9.3 and <= v1.9.6
- upgrade from v1.9.3 to v1.10.0:  there is 1 migration step >v1.9.3 and <= v1.10.0

The tricky part here is which version, because required changes are somehow linked to kubeproxy version, kubelet version and kubeadm version...